src/PufferVaultMainnet.sol

```diff
@@ -80,20 +105,107 @@ contract PufferVaultMainnet is PufferVault {
         }
 
         uint256 assets = previewRedeem(shares);
+
+        _checkDailyWithdrawalLimits(assets); //@todo figure if it is better to override `maxWithdraw`
+
+        _wrapETH(assets);
+
         // solhint-disable-next-line func-named-parameters
         _withdraw(_msgSender(), receiver, owner, assets, shares);
 
         return assets;
     }
 
-    // slither-disable-next-line dead-code
-    function _wrap(uint256 amount) internal {
-        _WETH.deposit{ value: amount }();
+    /**
+     * @param withdrawalAmount is the assets amount, not shares
+     */
+    function _checkDailyWithdrawalLimits(uint256 withdrawalAmount) internal {
+        VaultStorage storage $ = _getPufferVaultStorage();
+
+        // If daily withdrawal limit is 0, then there is no limit
+        if ($.dailyWithdrawalLimit == 0) {
+            return;
+        }
+
+        // Check if it's a new day to reset the withdrawal count
+        if ($.lastWithdrawalDay < block.timestamp / 1 days) {
+            $.lastWithdrawalDay = uint64(block.timestamp / 1 days);
+            $.withdrawnToday = 0;
+        }
+
+        if ($.withdrawnToday + withdrawalAmount > $.dailyWithdrawalLimit) {
+            revert ExceededDailyWithdrawalLimit($.dailyWithdrawalLimit, $.withdrawnToday, withdrawalAmount);
+        }
+
+        $.withdrawnToday += uint96(withdrawalAmount);
+    }
+
+    /**
+     * @notice Deposits native ETH
+     */
+    function depositETH(address receiver) public payable virtual returns (uint256) {
+        uint256 maxAssets = maxDeposit(receiver);
+        if (msg.value > maxAssets) {
+            revert ERC4626ExceededMaxDeposit(receiver, msg.value, maxAssets);
+        }
+
+        uint256 shares = previewDeposit(msg.value);
+        _mint(receiver, shares);
+        emit Deposit(_msgSender(), receiver, msg.value, shares);
+
+        return shares;
+    }
+
+    /**
+     * @notice Transfers ETH to a specified address
+     * @dev Restricted to PufferProtocol
+     * We use it to transfer ETH to PufferModule
+     * copied from https://github.com/Vectorized/solady/blob/fb11b3e9ea6c1aafdbd0a1515ff440509d60bff9/src/utils/SafeTransferLib.sol#L64
+     * @param to The address to transfer ETH to
+     * @param ethAmount The amount of ETH to transfer
+     */
+    function transferETH(address to, uint256 ethAmount) external restricted {
+        // Our Vault will hold ETH & WETH
+        // If we don't have enough ETH for the transfer, unwrap WETH
+        uint256 ethBalance = address(this).balance;
+        if (ethBalance < ethAmount) {
+            _WETH.withdraw(ethAmount - ethBalance);
+        }
+
+        /// @solidity memory-safe-assembly
```

Comment:
Suggestion to use normal .call() to transfer the amount

The assembly looks out of the place as we don't have that much assembly, and this function is not used as often. open to discuss